### ISSUE TITLE
Inplace update error in loading state_dict for YOLOV8 while implementing Federated Learning

### ISSUE THREAD

**Issue by shubhamjanu** [ultralytics/ultralytics#8804]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

I am facing an issue while implementing federated learning with yolov8. The code runs fine upto first round of training for clients. After one round, when parameters are received from the server, the set_parameter() method throws an error which goes as :
Traceback (most recent call last):
File "C:\Users\Administrator\Desktop\UTokyo\Japan51\client.py", line 34, in
fl.client.start_numpy_client(
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\app.py", line 500, in start_numpy_client
start_client(
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\app.py", line 248, in start_client
_start_client_internal(
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\app.py", line 382, in _start_client_internal
out_message = app(message=message, context=context)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\flower.py", line 76, in call
return self._call(message, context)
^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\flower.py", line 66, in ffn
out_message = handle_legacy_message_from_tasktype(
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\message_handler\message_handler.py", line 135, in handle_legacy_message_from_tasktype
fit_res = maybe_call_fit(
^^^^^^^^^^^^^^^
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\client.py", line 234, in maybe_call_fit
return client.fit(fit_ins)
^^^^^^^^^^^^^^^^^^^
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\flwr\client\numpy_client.py", line 238, in _fit
results = self.numpy_client.fit(parameters, ins.config) # type: ignore
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
File "C:\Users\Administrator\Desktop\UTokyo\Japan51\client.py", line 22, in fit
set_parameters(net, parameters)
File "C:\Users\Administrator\Desktop\UTokyo\Japan51\client.py", line 10, in set_parameters
net.load_state_dict(state_dict, strict=True)
File "C:\Users\Administrator\AppData\Local\Programs\Python\Python312\Lib\site-packages\torch\nn\modules\module.py", line 2153, in load_state_dict
raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for YOLO:
While copying the parameter named "model.model.0.conv.weight", whose dimensions in the model are torch.Size([48, 3, 3, 3]) and whose dimensions in the checkpoint are torch.Size([48, 3, 3, 3]), an exception occurred : ('Inplace update to inference tensor outside InferenceMode is not allowed.You can make a clone to get a normal tensor before doing inplace update.See https://github.com/pytorch/rfcs/pull/17 for more details.',).

Basically it seems like inplace update error, but i am not able to resolve it. I have tried many ways like making clone, deepcopy etc.
I am also attaching my code files: centralized.py, client.py and server.py as well as the error logs.
[centralized.txt](https://github.com/ultralytics/ultralytics/files/14544793/centralized.txt)
[client.txt](https://github.com/ultralytics/ultralytics/files/14544794/client.txt)
[error.txt](https://github.com/ultralytics/ultralytics/files/14544795/error.txt)
[server.txt](https://github.com/ultralytics/ultralytics/files/14544796/server.txt)

Please help, i am stuck at this issue since many days. If you need any additional info., let me know.

**Comment by glenn-jocher** [comment]
@shubhamjanu hello! Thanks for reaching out and providing detailed information about the issue you're encountering. 🌟

From the error message, it seems like the problem arises due to an in-place update attempt on a tensor that's not allowed outside of PyTorch's InferenceMode. This can sometimes happen when loading state dictionaries, especially in complex setups like federated learning.

One potential workaround to try is explicitly cloning the parameter tensors from the received state dictionary before loading them into your model. This can be done by iterating over the state dictionary and cloning each tensor. Here's a quick example of how you might modify your `set_parameters` function:

```python
def set_parameters(model, parameters):
    new_state_dict = {}
    for key, param in parameters.items():
        new_state_dict[key] = param.clone()  # Clone the parameters
    model.load_state_dict(new_state_dict, strict=True)
```

This approach ensures that you're not attempting an in-place update on the original tensors received from the server but rather on a cloned version that's safe to modify.

If you've already tried cloning and the issue persists, it might be helpful to double-check the PyTorch version compatibility and ensure that all tensors are properly detached and cloned before the update.

I hope this helps! If the issue continues, please feel free to share more details or any updates on your attempts. We're here to help. Good luck! 🍀

**Comment by shubhamjanu** [comment]
@glenn-jocher The parameters received from server are in list format, so I had to match keys and then convert those into dictionary. 
def set_parameters(net, parameters):
    params_dict = zip(net.state_dict().keys(), parameters)
    state_dict = OrderedDict({k: torch.Tensor(v).clone() for k, v in params_dict})
    new_state_dict = OrderedDict()
    for key, param in state_dict.items():
        new_state_dict[key] = param.clone()  # Clone the parameters
    net.load_state_dict(new_state_dict, strict=True)
    
    But this code also did not work. It gave same inplace update error.
    
Then, I removed evaluate() method completely from my client.py code. The code started working. Now, I am able to train my model for more than one round. After every round of training of yolov8 model, there is inherent self validation of the model with best weights. Therefore, I think I can do away with evaluate() method. What is your opinion?

**Comment by glenn-jocher** [comment]
@shubhamjanu hey there! 😊 It's great to see you're making progress with your federated learning implementation using YOLOv8. It looks like you've found a workaround by removing the `evaluate()` method, which is interesting!

Regarding the inplace update error, your approach to converting the parameters list to a dictionary and cloning them seems correct. This error can be quite tricky, and it's good that you've found a temporary solution by adjusting your code flow.

As for doing away with the `evaluate()` method, if your training and validation needs are being met without it, and you're able to proceed with multiple rounds of training successfully, it sounds like a practical solution for your specific use case. YOLOv8 does indeed perform validation checks with the best weights, which might suffice for your project's requirements.

However, keep in mind that explicit evaluation can sometimes offer more control and insights, especially for fine-tuning or debugging. If you find yourself needing more detailed evaluation metrics or want to implement custom validation logic in the future, you might consider reintroducing it in a way that doesn't lead to the inplace update issue.

Happy coding, and best of luck with your federated learning project! If you have more questions or hit another roadblock, feel free to reach out. 🚀

**Comment by shubhamjanu** [comment]
@glenn-jocher Hi! I wish to report one strange observation. I was able to run my federated code with available CPU. Today, I installed CUDA 11.8 and tried to run the code. It is getting stuck and not able to start training. It is also trying "AMP: running Automatic Mixed Precision (AMP) checks with YOLOv8n...and Downloading https://github.com/ultralytics/assets/releases/download/v8.1.0/yolov8n.pt to 'yolov8n.pt'...". Which was not observed when training with CPU.
To reproduce the issue, I again tried running with CPU only and it was working fine.
I have installed cuda and cudaCNN compatible with pytorch after checking on official pytorch website and they were installed successfully without any warning or error message. I am using an AWS instance windows XP 2022 server.
I am attaching my code files and detailed error log. The training does not start and then grpc channel closes after waiting for sometime.
[centralized.txt](https://github.com/ultralytics/ultralytics/files/14557623/centralized.txt)
[client.txt](https://github.com/ultralytics/ultralytics/files/14557624/client.txt)
[error.txt](https://github.com/ultralytics/ultralytics/files/14557625/error.txt)
[server.txt](https://github.com/ultralytics/ultralytics/files/14557626/server.txt)

Please guide how to resolve above issue.

**Comment by glenn-jocher** [comment]
Hi @shubhamjanu! 👋 It sounds like you're encountering an interesting situation with the transition from CPU to GPU training using CUDA. Let's address the issue step by step. 🛠️

When switching to CUDA, it’s good practice to ensure that your system's CUDA and cuDNN versions are compatible with the version of PyTorch you're using. It looks like you’ve checked compatibility, which is great!

The behavior you're seeing, including the AMP checks and the model download, is expected on the first run with a GPU. AMP and model downloading should not fundamentally cause training to get stuck though.

Since the code works fine on CPU but gets stuck on GPU, this might indicate an environment setup issue related to CUDA. Here are a few things you might consider checking:

1. **CUDA Path**: Ensure CUDA's bin directories are properly added to your system's `PATH`.

2. **PyTorch and CUDA Compatibility**: Double-check the PyTorch to CUDA compatibility. Even though you've checked, a misalignment here can cause issues.

3. **CUDA Installation Verification**: Run a simple CUDA sample program (like `deviceQuery` from the CUDA samples) to make sure CUDA is correctly installed and the GPU is accessible.

4. **PyTorch Test**: Run a simple PyTorch script to verify if PyTorch can utilize your GPU. Something like this could help:
   ```python
   import torch
   print(torch.cuda.is_available())
   print(torch.cuda.get_device_name(0))
   ```
   This verifies whether PyTorch can detect and use the GPU.

Since the issue arises when utilizing CUDA and involves AMP checks, you might also try disabling AMP temporarily to see if it affects the stuck training process. While not a solution, it can be a useful diagnostic step.

If none of these steps reveal the issue, considering the complexity of federated learning and potential nuances of running on an AWS Windows instance with an NVIDIA GPU, it could be a more intricate compatibility issue.

I hope these suggestions help get you moving in the right direction. If the problem persists, could you share the versions of PyTorch, CUDA, and cuDNN you're using? It might offer further insights into the issue. Keep us posted on your progress!
